client/internal/receiver.go

```diff
@@ -0,0 +1,167 @@
+package internal
+
+import (
+	"context"
+
+	"github.com/gorilla/websocket"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+// Receiver implements the client's receiving portion of OpAMP protocol.
+type Receiver struct {
+	conn       *websocket.Conn
+	logger     types.Logger
+	sender     *Sender
+	runContext context.Context
+	callbacks  types.Callbacks
+}
+
+func NewReceiver(logger types.Logger, callbacks types.Callbacks, conn *websocket.Conn, sender *Sender) *Receiver {
+	return &Receiver{
+		conn:      conn,
+		logger:    logger,
+		sender:    sender,
+		callbacks: callbacks,
+	}
+}
+
+func (r *Receiver) ReceiverLoop(ctx context.Context) {
+	runContext, cancelFunc := context.WithCancel(ctx)
+	r.runContext = runContext
+
+out:
+	for {
+		var message protobufs.ServerToAgent
+		if err := r.receiveMessage(&message); err != nil {
+			if ctx.Err() == nil && !websocket.IsCloseError(err, websocket.CloseNormalClosure) {
+				r.logger.Errorf("Unexpected error while receiving: %v", err)
+			}
+			break out
+		} else {
+			r.processReceivedMessage(&message)
+		}
+	}
+
+	cancelFunc()
+}
+
+func (r *Receiver) receiveMessage(msg *protobufs.ServerToAgent) error {
+	_, bytes, err := r.conn.ReadMessage()
+	if err != nil {
+		return err
+	}
+	err = proto.Unmarshal(bytes, msg)
+	if err != nil {
+		r.logger.Errorf("Cannot decode received message: %v", err)
+	}
+	return err
+}
+
+func (r *Receiver) processReceivedMessage(p *protobufs.ServerToAgent) {
+	data := p.GetDataForAgent()
+	if data != nil {
+		r.rcvDataForAgent(data)
+		return
+	}
+
+	err := p.GetErrorResponse()
+	if err != nil {
+		r.processErrorResponse(err)
+	}
+}
+
+func (r *Receiver) rcvDataForAgent(data *protobufs.DataForAgent) {
+	if r.callbacks != nil {
+		reportStatus := false
+
+		reportStatus = r.rcvRemoteConfig(data.RemoteConfig) || reportStatus
+		reportStatus = r.rcvConnectionSettings(data.ConnectionSettings) || reportStatus
```

Comment:
`rcvConnectionSetting` will never return `true` though, will it?

Reply:
Correct. Removed.